src/add.cpp

```diff
@@ -330,8 +330,9 @@ RCP<const Basic> sub(const RCP<const Basic> &a, const RCP<const Basic> &b)
     return add(a, mul(minus_one, b));
 }
 
-RCP<const Basic> add_expand(const RCP<const Add> &self)
+RCP<const Basic> add_expand(const RCP<const Basic> &self_)
 {
+    const RCP<const Add> self = rcp_static_cast<const Add>(self_);
```

Comment:
@isuruf I pushed your implementation into my PR. But I think there is significant overhead to `std::function` and `std::bind`, so we might need to figure out an alternative. But it seems to work.